src/frequenz/sdk/power/_distribution_algorithm.py

```diff
@@ -336,57 +346,98 @@ def _distribute_power(
             # ratio = 0, means all remaining batteries reach max SoC lvl or have no
             # capacity
             if is_close_to_zero(ratio):
                 distribution[inverter.component_id] = 0.0
                 continue
 
-            distribution[inverter.component_id] = (
-                power_to_distribute * battery_ratio / ratio
-            )
-
+            power_to_distribute = power_w - reserved_power
+            calculated_power = power_to_distribute * battery_ratio / ratio
+            reserved_power += max(calculated_power, excl_bound)
             used_ratio += battery_ratio
-
+            ratio = sum_ratio - used_ratio
             # If the power allocated for that inverter is out of bound,
             # then we need to distribute more power over all remaining batteries.
-            upper_bound = upper_bounds[inverter.component_id]
-            if distribution[inverter.component_id] > upper_bound:
-                distribution[inverter.component_id] = upper_bound
-                distributed_power += upper_bound
-                # Distribute only the remaining power.
-                power_to_distribute = power_w - distributed_power
-                # Distribute between remaining batteries
-                ratio = sum_ratio - used_ratio
+            incl_bound = incl_bounds[inverter.component_id]
+            if calculated_power > incl_bound:
+                excess_reserved[inverter.component_id] = incl_bound - excl_bound
+                # # Distribute between remaining batteries
+            elif calculated_power < excl_bound:
+                deficits[inverter.component_id] = calculated_power - excl_bound
             else:
-                distributed_power += distribution[inverter.component_id]
+                excess_reserved[inverter.component_id] = calculated_power - excl_bound
+
+            distributed_power += excl_bound
+            distribution[inverter.component_id] = excl_bound
+
+        for inverter_id, deficit in deficits.items():
+            while not math.isclose(deficit, 0.0, abs_tol=1e-6) and deficit < 0.0:
+                take_from = max(excess_reserved.items(), key=lambda item: item[1])
+                if math.isclose(take_from[1], 0.0, abs_tol=1e-6) or take_from[1] < 0.0:
+                    break
+                if take_from[1] >= -deficit or math.isclose(
+                    take_from[1], -deficit, abs_tol=1e-6
+                ):
+                    excess_reserved[take_from[0]] += deficit
+                    deficits[inverter_id] = 0.0
+                    deficit = 0.0
+                else:
+                    deficit += excess_reserved[take_from[0]]
+                    deficits[inverter_id] = deficit
+                    excess_reserved[take_from[0]] = 0.0
+
+        for inverter_id, excess in excess_reserved.items():
+            distribution[inverter_id] += excess
+            distributed_power += excess
+
+        for inverter_id, deficit in deficits.items():
```

Comment:
This duplicated for-loop is not needed as its content could be added in the first loop for deficits, right after the end of the `while`-loop